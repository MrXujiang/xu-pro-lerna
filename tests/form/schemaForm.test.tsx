import React from 'react';
import { mount, render } from 'enzyme';
import { BetaSchemaForm } from '@ant-design/pro-form';
import type { ProFormColumnsType } from '@ant-design/pro-form';
import { waitForComponentToPaint } from '../util';
import { Input } from 'antd';
import { act } from 'react-dom/test-utils';

const columns: ProFormColumnsType<any>[] = [
  {
    title: '序号',
    dataIndex: 'index',
    valueType: 'indexBorder',
  },
  {
    title: '标题',
    dataIndex: 'title',
    width: 200,
  },
  {
    title: (_, type) => (type === 'table' ? '状态' : '列表状态'),
    dataIndex: 'state',
    initialValue: 'all',
    valueType: 'select',
    valueEnum: {
      all: { text: '全部', status: 'Default' },
      open: {
        text: '未解决',
        status: 'Error',
      },
      closed: {
        text: '已解决',
        status: 'Success',
      },
    },
  },
  {
    title: '排序方式',
    key: 'direction',
    dataIndex: 'direction',
    valueType: 'select',
    valueEnum: {
      asc: '正序',
      desc: '倒序',
    },
  },
  {
    title: '创建时间',
    key: 'since',
    dataIndex: 'created_at',
    valueType: 'dateTime',
  },
  {
    title: 'option',
    valueType: 'option',
    dataIndex: 'id',
  },
];

describe('SchemaForm', () => {
  it('😊 SchemaForm support columns', () => {
    const html = render(<BetaSchemaForm columns={columns} />);
    expect(html).toMatchSnapshot();
  });

  it('😊 SchemaForm support dependencies', async () => {
    const onChange = jest.fn();
    const html = mount(
      <BetaSchemaForm
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            width: 200,
            initialValue: 'name',
            fieldProps: {
              id: 'title',
            },
          },
          {
            title: '选择器',
            dataIndex: 'state',
            valueType: 'select',
            dependencies: ['title'],
            request: async ({ title }) => {
              onChange(title);
              return [
                {
                  label: title,
                  value: 'title',
                },
              ];
            },
          },
        ]}
      />,
    );
    expect(onChange).toBeCalledWith('name');
    act(() => {
      html.find('input#title').simulate('change', {
        target: {
          value: 'qixian',
        },
      });
    });
    await waitForComponentToPaint(html);
    expect(onChange).toBeCalledWith('qixian');
  });

  it('🐲 SchemaForm support StepsForm', async () => {
    const html = mount(
      <BetaSchemaForm
        layoutType="StepsForm"
        steps={[
          {
            title: '表单1',
          },
          {
            title: '表单2',
          },
          {
            title: '表单3',
          },
        ]}
        columns={[
          [
            {
              title: '邮件',
              dataIndex: 'email',
            },
          ],
          [
            {
              title: '姓名',
              dataIndex: 'name',
            },
          ],
          [
            {
              title: '地址',
              dataIndex: 'addr',
            },
          ],
        ]}
      />,
    );
    await waitForComponentToPaint(html);
    expect(html.find('span.ant-steps-icon').length).toBe(3);
    expect(html.find('div.ant-steps-item-title').at(0).text()).toBe('表单1');
    expect(html.find('div.ant-steps-item-title').at(1).text()).toBe('表单2');
    expect(html.find('div.ant-steps-item-title').at(2).text()).toBe('表单3');
    await waitForComponentToPaint(html, 100);
    act(() => {
      html.unmount();
    });
  });

  it('😊 SchemaForm support table columns', async () => {
    const html = mount(<BetaSchemaForm columns={columns} />);
    await waitForComponentToPaint(html);
    expect(html.find('div.ant-form-item').length).toBe(4);
  });

  it('😊 SchemaForm support render', async () => {
    const html = mount(
      <BetaSchemaForm
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            readonly: true,
            width: 200,
            render: () => {
              return <Input id="test" />;
            },
          },
        ]}
      />,
    );
    await waitForComponentToPaint(html);
    expect(html.find('#test').exists()).toBeTruthy();
  });

  it('😊 SchemaForm support render', async () => {
    const html = mount(
      <BetaSchemaForm
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            width: 200,
            renderFormItem: () => {
              return <Input id="test" />;
            },
          },
        ]}
      />,
    );
    await waitForComponentToPaint(html);
    expect(html.find('#test').exists()).toBeTruthy();
  });

  it('😊 SchemaForm support render', async () => {
    const html = mount(
      <BetaSchemaForm
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            width: 200,
            renderFormItem: (_, { defaultRender }) => {
              return defaultRender(_);
            },
          },
        ]}
      />,
    );
    await waitForComponentToPaint(html);
    expect(html.find('input').exists()).toBeTruthy();
  });

  it('😊 SchemaForm support hidenInForm', async () => {
    const html = mount(
      <BetaSchemaForm
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            width: 200,
            renderFormItem: () => {
              return <Input id="title" />;
            },
          },
          {
            title: '类型',
            dataIndex: 'category',
            width: 200,
            hideInForm: true,
            renderFormItem: () => {
              return <Input id="category" />;
            },
          },
        ]}
      />,
    );
    await waitForComponentToPaint(html);
    expect(html.find('#title').exists()).toBeTruthy();
    expect(html.find('#category').exists()).toBeFalsy();
  });
});
