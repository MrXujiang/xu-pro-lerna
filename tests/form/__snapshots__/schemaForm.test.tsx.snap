// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaForm üòä SchemaForm support columns 1`] = `
<form
  autocomplete="off"
  class="ant-form ant-form-vertical"
>
  <input
    style="display:none"
    type="text"
  />
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="title"
        title="Ê†áÈ¢ò"
      >
        Ê†áÈ¢ò
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <span
            class="ant-input-affix-wrapper"
            style="width:200px"
          >
            <input
              class="ant-input"
              id="title"
              placeholder="ËØ∑ËæìÂÖ•"
              type="text"
              value=""
            />
            <span
              class="ant-input-suffix"
            >
              <span
                aria-label="close-circle"
                class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                role="button"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="state"
        title="ÂàóË°®Áä∂ÊÄÅ"
      >
        ÂàóË°®Áä∂ÊÄÅ
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-select ant-pro-filed-search-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
            style="min-width:100px"
          >
            <div
              class="ant-select-selector"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-activedescendant="state_list_0"
                  aria-autocomplete="list"
                  aria-controls="state_list"
                  aria-haspopup="listbox"
                  aria-owns="state_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="state"
                  readonly=""
                  role="combobox"
                  style="opacity:0"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
                title="ÂÖ®ÈÉ®"
              >
                ÂÖ®ÈÉ®
              </span>
            </div>
            <span
              aria-hidden="true"
              class="ant-select-arrow"
              style="user-select:none;-webkit-user-select:none"
              unselectable="on"
            >
              <span
                aria-label="down"
                class="anticon anticon-down ant-select-suffix"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </span>
            </span>
            <span
              aria-hidden="true"
              class="ant-select-clear"
              style="user-select:none;-webkit-user-select:none"
              unselectable="on"
            >
              <span
                aria-label="close-circle"
                class="anticon anticon-close-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="direction"
        title="ÊéíÂ∫èÊñπÂºè"
      >
        ÊéíÂ∫èÊñπÂºè
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-select ant-pro-filed-search-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
            style="min-width:100px"
          >
            <div
              class="ant-select-selector"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-activedescendant="direction_list_0"
                  aria-autocomplete="list"
                  aria-controls="direction_list"
                  aria-haspopup="listbox"
                  aria-owns="direction_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="direction"
                  readonly=""
                  role="combobox"
                  style="opacity:0"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-placeholder"
              >
                ËØ∑ÈÄâÊã©
              </span>
            </div>
            <span
              aria-hidden="true"
              class="ant-select-arrow"
              style="user-select:none;-webkit-user-select:none"
              unselectable="on"
            >
              <span
                aria-label="down"
                class="anticon anticon-down ant-select-suffix"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="since"
        title="ÂàõÂª∫Êó∂Èó¥"
      >
        ÂàõÂª∫Êó∂Èó¥
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-picker"
          >
            <div
              class="ant-picker-input"
            >
              <input
                autocomplete="off"
                id="since"
                placeholder="ËØ∑ÈÄâÊã©"
                readonly=""
                size="21"
                title=""
                value=""
              />
              <span
                class="ant-picker-suffix"
              >
                <span
                  aria-label="calendar"
                  class="anticon anticon-calendar"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="calendar"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right:8px"
    >
      <button
        class="ant-btn ant-btn-primary"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
    <div
      class="ant-space-item"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
    </div>
  </div>
</form>
`;
